### PR TITLE
Add support for paths in tsconfig.json and jsconfig.json

### LINKED ISSUE #7779: Automatically configure webpack to handle `tsconfig.json`'s `baseUrl` and `paths`
# Bug report

## Describe the bug

Reproduction here: https://github.com/liamdanielduffy/next-9-baseurl-reproduction

I've setup a simple Next.js project using v 9.0.0, with a tsconfig.json generated by Next

I've added the 'baseUrl' compiler option to 'tsconfig.json' in the root of my repository. The baseUrl is `./` since I want paths in my Next app to resolve relative to the root of my repository. 

I've created a test import in `pages/index.tsx` which imports a string constant.

When I run this locally with `now dev`, I get the following error:

`ModuleNotFoundError: Module not found: Error: Can't resolve 'constants/test'`

I've also deployed this to Now under this project: `/liamdanielduffy/app-workflow/6ijvjnt5c`

I get the same error during build:

```
./pages/index.tsx
Module not found: Can't resolve 'constants/test' in '/tmp/28475d91/pages'
```

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Clone the provided repository
2. Run `now dev` in the root to see the local build error
3. Run `now` to deploy and see the build error in the Now environment

## Expected behavior

Next should be able to resolve modules imported within `.ts` or `.tsx` files relative to the `"baseUrl"` set in `"compilerOptions"` of tsconfig.json

### PR DESCRIPTION
**Pull request by timneutkens**
Adds support for `"paths"` in both `tsconfig.json` or `jsconfig.json`.

Examples:

**Single module alias**

```json
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "mytheme": ["components/mytheme.js"]
    }
  }
}
```

**Wildcard alias**

```json
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@c/*": ["components/*"],
    }
  }
}
```

**Fallback behavior**

```json
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@mytheme/*": ["themes/theme-a/*", "lib/theme-b/*"]
    }
  }
}
```

Fixes #7779

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| buildDuration | 10.7s | 10.5s | -186ms |
| nodeModulesSize | 52.8 MB | 52.8 MB | ⚠️ +14.6 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.24 kB | 6.24 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..c6c1.js gzip | 10.1 kB | 10.1 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.2 kB | 56.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.78 kB | 4.78 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.71 kB | 6.71 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.4 kB | 51.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| _app.js gzip | 1.24 kB | 1.24 kB | ✓ |
| _error.js gzip | 3.15 kB | 3.15 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.03 kB | 2.03 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.86 kB | 7.86 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| _app.module.js gzip | 594 B | 594 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.28 kB | 5.28 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| index.html gzip | 916 B | 916 B | ✓ |
| link.html gzip | 924 B | 924 B | ✓ |
| withRouter.html gzip | 913 B | 913 B | ✓ |
| Overall change | 2.75 kB | 2.75 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| buildDuration | 11.6s | 11.3s | -324ms |
| nodeModulesSize | 52.8 MB | 52.8 MB | ⚠️ +14.6 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.24 kB | 6.24 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..c6c1.js gzip | 10.1 kB | 10.1 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.2 kB | 56.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.78 kB | 4.78 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.71 kB | 6.71 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.4 kB | 51.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| _app.js gzip | 1.24 kB | 1.24 kB | ✓ |
| _error.js gzip | 3.15 kB | 3.15 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.03 kB | 2.03 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.86 kB | 7.86 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| _app.module.js gzip | 594 B | 594 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.28 kB | 5.28 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong> Overall decrease ✓</summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| _error.js gzip | 293 kB | 293 kB | -111 B |
| 404.html gzip | 1.32 kB | 1.32 kB | ✓ |
| hooks.html gzip | 956 B | 956 B | ✓ |
| index.js gzip | 294 kB | 294 kB | ⚠️ +7 B |
| link.js gzip | 301 kB | 301 kB | -34 B |
| routerDirect.js gzip | 300 kB | 300 kB | ⚠️ +149 B |
| withRouter.js gzip | 300 kB | 300 kB | -129 B |
| Overall change | 1.49 MB | 1.49 MB | -118 B |

</details>

</details>

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| buildDuration | 9.5s | 9.5s | ⚠️ +55ms |
| nodeModulesSize | 52.8 MB | 52.8 MB | ⚠️ +14.6 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.24 kB | 6.24 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..c6c1.js gzip | 10.1 kB | 10.1 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.2 kB | 56.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.78 kB | 4.78 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.71 kB | 6.71 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.4 kB | 51.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| _app.js gzip | 1.24 kB | 1.24 kB | ✓ |
| _error.js gzip | 3.15 kB | 3.15 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.03 kB | 2.03 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.86 kB | 7.86 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| _app.module.js gzip | 594 B | 594 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.28 kB | 5.28 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| index.html gzip | 916 B | 916 B | ✓ |
| link.html gzip | 924 B | 924 B | ✓ |
| withRouter.html gzip | 913 B | 913 B | ✓ |
| Overall change | 2.75 kB | 2.75 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| buildDuration | 10.2s | 10.5s | ⚠️ +249ms |
| nodeModulesSize | 52.8 MB | 52.8 MB | ⚠️ +14.6 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.24 kB | 6.24 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..c6c1.js gzip | 10.1 kB | 10.1 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.2 kB | 56.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.78 kB | 4.78 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.71 kB | 6.71 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.4 kB | 51.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| _app.js gzip | 1.24 kB | 1.24 kB | ✓ |
| _error.js gzip | 3.15 kB | 3.15 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.03 kB | 2.03 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.86 kB | 7.86 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| _app.module.js gzip | 594 B | 594 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.28 kB | 5.28 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | timneutkens/next.js add/jsconfig-paths | Change |
| - | - | - | - |
| _error.js gzip | 293 kB | 294 kB | ⚠️ +705 B |
| 404.html gzip | 1.32 kB | 1.32 kB | ✓ |
| hooks.html gzip | 956 B | 956 B | ✓ |
| index.js gzip | 293 kB | 294 kB | ⚠️ +631 B |
| link.js gzip | 301 kB | 301 kB | -311 B |
| routerDirect.js gzip | 300 kB | 300 kB | -177 B |
| withRouter.js gzip | 300 kB | 300 kB | ⚠️ +114 B |
| Overall change | 1.49 MB | 1.49 MB | ⚠️ +962 B |

</details>

</details>

**Comment by Janpot** [comment]
@timneutkens Since `paths` is read from json files, the value is completely static. I'd be interested in being able to set these from `next.config.js`. That way they can also be populated from environment variables (so that they can be populated from e.g. ZEIT now `buildEnv`). Is that a use-case that could be considered for this PR as well?

**Comment by timneutkens** [comment]
> That way they can also be populated from environment variables (so that they can be populated from e.g. ZEIT now buildEnv). Is that a use-case that could be considered for this PR as well?

That's not supported by most code editors while tsconfig and jsconfig are. What's the use case?

**Comment by timneutkens** [comment]
Allowing dynamic values also makes it significantly harder to do any type of caching in the future.

**Review by Timer — APPROVED** [review]
LGTM! Excited for this existing convention based feature to work out of the box with zero added configuration to Next.js. 😌

**Comment by JerryGreen** [comment]
Wow, I was waiting for this feature! I like using `~/` alias for my src folder, and had to setup this in both next.config.js and tsconfig.json. Now one setting is enough. Thanks a lot!

**Comment by Janpot** [comment]
> That's not supported by most code editors while tsconfig and jsconfig are. What's the use case?

@timneutkens [Use case](https://github.com/Janpot/nymus/tree/master/examples/next.js#example) is to create a localized next.js app, where each locale is an instance of the app on its own. locale data is made dynamic by [configuring an alias](https://github.com/Janpot/nymus/blob/c0c7e9048f3dbabb175907626911c4fe1f882b7f/examples/next.js/next.config.js#L34) to a locale folder. (This is just a POC, still trying to feel the water on where next.js is going to go feature-wise.)

**Comment by Janpot** [comment]
Just a thought, but can this be (ab)used to enable "transpile node_modules"? as in 

```json
"paths": {
  "my-module/*": "node_modules/my-module/*"
}
```

**Comment by felixmosh** [comment]
I'm using [`tsconfig-paths-webpack-plugin`](https://www.npmjs.com/package/tsconfig-paths-webpack-plugin) for more then a year, and it just works.

Do you know what are differences between your implementation and this lib?

**Comment by timneutkens** [comment]
> I'm using [`tsconfig-paths-webpack-plugin`](https://www.npmjs.com/package/tsconfig-paths-webpack-plugin) for more then a year, and it just works.
> 
> Do you know what are differences between your implementation and this lib?

Significantly more overhead as it uses another library that implements filesystem calls itself instead of relying on webpack.

Besides that it didn't seem to be based on the actual TypeScript source code, meaning there could be differences in how the output path is calculated.

It also maintained a lot of baggage, eg webpack 3 support and checks that are not needed as we know exactly how the plugin will execute in Next.js.

**Comment by IljaDaderko** [comment]
Is this supporting monorepo structures where paths are defined at repo root? i.e:

```
packages/
  nextApp/
     tsconfig.json
tsconfig.json
```

I added: `jsconfigPaths: true` to my config, installed `9.3.2-canary.3` and removed webpack alias I used to have, but am getting errors that can't resolve paths unless I add those webpack aliases back.

**Comment by timneutkens** [comment]
Didn't add resolving of `extends` currently

**Comment by Janpot** [comment]
You probably considered this already but mentioning it to make sure. Another option is to build it on top of the typescript compiler API. It seems to provide a [`host.resolveModuleName`](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API#customizing-module-resolution).

**Comment by Favna** [comment]
@timneutkens I just upgraded my NextJS which includes this change and removed `tsconfig-paths-webpack-plugin` and tried building my code, but to no avail, webpack just errors out:
```
$ next build "src/app"
Creating an optimized production build


./pages/index.tsx
Module not found: Can't resolve '@Config/HomeProjectCardData' in 'E:\dev\website\src\app\pages'


> Build error occurred
Error: > Build failed because of webpack errors
    at build (E:\dev\website\node_modules\next\dist\build\index.js:13:900)
```

I've branched my code, could you have a look what's going wrong here? My app is based on the Firebase sample.

Branch (tsconfig specifically): https://github.com/favware/website/blob/pluginless-paths/src/app/tsconfig.json

**Comment by timneutkens** [comment]
You forgot to add the experimental flag in next.config.js

**Comment by Favna** [comment]
Ah I missed that yeah, thanks!

**Comment by vvo** [comment]
Hey @timneutkens it seems I can now use this without the experimental flag, do you confirm? If so, should we add documentation for it? Thanks!

**Comment by timneutkens** [comment]
> Hey @timneutkens it seems I can now use this without the experimental flag, do you confirm? If so, should we add documentation for it? Thanks!

We don't document experimental flags as they are experimental and can be removed at any time in some cases (this one will land soon).

**Comment by 4ortytwo** [comment]
Hi everyone! My issue is that when I use a component first and then import it from Material UI, the component is imported from '@r(root)/node_modules/@materialui-core...etc'. Is there a way around it?

**Comment by timneutkens** [comment]
@4ortytwo can you provide a clear example? It's hard to follow what you're asking.

**Comment by 4ortytwo** [comment]
I have this in my tsconfig 
```js
 "baseUrl": ".",
    "paths": {
      "@r/*": ["./*"], // alias for root
      "@s/*": ["./src/*"] // alias for src
    }
```

And this in my next.config 

```
experimental: {
    jsconfigPaths: true, 
  },
```

Let's say I created a new tsx file and used Material-ui ``` <Container></Container> ``` in it without prior import. I see that it's not imported and use 'Add import statement' of WebStorm. 
At the top, I see that it's imported from  ```@r(root)/node_modules/@materialui-core...etc```. 
Anything imported from any other library also has a similarly looking path. 

I wonder whether it could be related to the IDE. I'd like to be able to use aliases for my paths but I still want library imports to look concise. How can I achieve that?

**Comment by Favna** [comment]
@4ortytwo That's just how TypeScript resolves paths and in a way it is sort of related to your IDE but I doubt using something like VSCode is going to change it much. My own recommendation would be to not have a path alias for root dir and instead just keep files in a `src` dir (which you can then also set as the `rootDir` and `baseDir` in the tsconfig, which will somewhat improve TSServer since it doesn't have to traverse as many files).

Alternatively, paths are resolved bottom to top (with top taking higher priority I mean) so if you really need to keep the root path alias then add another alias, not sure if this exactly works, might need some tweaking. But I *think* it'll work:
```json
{
    "paths": {
      "@material-ui/*": ["./node_modules/@material-ui/*"],
      "@r/*": ["./*"],
      "@s/*": ["./src/*"]
    }
}
```

---

By example in another project where I work we have our paths set up like this:
```json
{
		"paths": {
			"@utils/*": ["lib/util/*"],
			"@lib/*": ["lib/*"],
			"@root/*": ["*"]
		}
}
```

And whenever a file is in `lib/util` it will suggest `@utils/file.ts` rather than `@lib/util/file.ts`.

**Comment by davecardwell** [comment]
> Didn't add resolving of `extends` currently

@timneutkens do you plan to support this? If so, is there an issue I can follow? I would definitely be interested in this feature but cannot use jsconfigPaths as it stands.

**Comment by Timer** [comment]
@davecardwell it's on the roadmap, but there's no ETA. Feel free to send a pull request adding support!

**Comment by HeadFox** [comment]
Hi everyone ✋ 
I spend a loooottt of time on this issue and still can't get it work.
There is no problem if the alias concern a folder outside of src but if it's in src, it's not working... .
I have the Typescript error in dev mode.
image

I am trying to do `import ExampleContainer from '@containers/ExampleContainer` where ExempleContainer is located at `src/containers/ExampleContainer`

Here is my tsconfig.json:
```json
{
  "compilerOptions": {
    "target": "es5",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": false,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "baseUrl": "./src",
    "paths": {
      "@containers/*": ["containers/*"],
      "@context/*": ["config/context/*"],
      "@components/*": ["components/*"],
      "@styles/*": ["styles/*"],
      "@contents/*": ["contents/*"],
      "@assets/*": ["assets/*"],
      "@utils/*": ["utils/*"],
      "@constants/*": ["config/constants/*"]
    }
  },
  "exclude": [
    "node_modules"
  ],
  "include": [
    "next-env.d.ts",
    "**/*.ts",
    "**/*.tsx"
  ]
}
```

My next.config.js :
```js
module.exports = {
  experimental: {
    jsconfigPaths: true,
  },
};
```

Thank you for your help 👍

**Comment by timneutkens** [comment]
Can you create a minimal repository?

**Comment by HeadFox** [comment]
Sure !
https://github.com/HeadFox/react-app-starter/tree/feature/migrate-to-typescript

**Comment by timneutkens** [comment]
Looks like it's related to yarn PnP because when you get rid of `.yarn` and `yarnrc.yml` it works fine

**Comment by HeadFox** [comment]
Okey thanks a lot👍 
I will ask @arcanis if he can help me on this.

**Comment by timneutkens** [comment]
I think we might have to run our plugin before the pnp resolver plugin, though I'm not sure how to test this and it seems incredibly hard to debug installed modules when using PnP.

**Comment by timneutkens** [comment]
Hopefully this fixes it, as said I couldn't test it: https://github.com/zeit/next.js/pull/12104

**Comment by HeadFox** [comment]
Okey thanks I will test it as soon as I can !
I also send a message in the Yarn's discord

**Comment by arcanis** [comment]
It's probably worth a separate issue, but generally speaking we don't recommend using paths or aliases (even though aliases work fine with Webpack last time I checked).

Instead, we strongly suggest to use the `link:` protocol, which is the idiomatic way to point a package name to a location in a way that all tools can understand without any configuration. More details [here](https://yarnpkg.com/features/protocols#why-is-the-link-protocol-recommended-over-aliases-for-path-mapping).

**Comment by lrosa007** [comment]
@arcanis worked great thanks

**Comment by lauralouiset** [comment]
This is feature is working for me, but eslint is throwing a "unable to resolve path to module" errors whenever I try this. The build looks fine, but eslint is unhappy. Any insight on how to fix this? Will #11986 help?

**Comment by vuthpov** [comment]
> module.exports = {
>   experimental: {
>     jsconfigPaths: true, // enables it for both jsconfig.json and tsconfig.json
>   }
> }

Hi, Im new to webpack and next.js.
How do add the flag above if my next.config.js look like this

`module.exports = withFonts(
    withCSS(
        withLess({
            webpack(config) {
                config.module.rules.push({
                    test: /\.svg(\?v=\d+\.\d+\.\d+)?$/,
                    use: [
                        {
                            loader: 'babel-loader',
                        },
                        {
                            loader: '@svgr/webpack',
                            options: {
                                babel: false,
                                icon: true,
                            },
                        },
                    ],
                });
                return config;
            },
        }),
    ),
);`

**Comment by vuthpov** [comment]
> Adds support for `"paths"` in both `tsconfig.json` or `jsconfig.json`.
> 
> Examples:
> 
> **Single module alias**
> 
> ```json
> {
>   "compilerOptions": {
>     "baseUrl": ".",
>     "paths": {
>       "mytheme": ["components/mytheme.js"]
>     }
>   }
> }
> ```
> 
> **Wildcard alias**
> 
> ```json
> {
>   "compilerOptions": {
>     "baseUrl": ".",
>     "paths": {
>       "@c/*": ["components/*"],
>     }
>   }
> }
> ```
> 
> **Fallback behavior**
> 
> ```json
> {
>   "compilerOptions": {
>     "baseUrl": ".",
>     "paths": {
>       "@mytheme/*": ["themes/theme-a/*", "lib/theme-b/*"]
>     }
>   }
> }
> ```
> 
> Temporarily behind an experimental feature flag:
> 
> ```js
> // next.config.js
> module.exports = {
>   experimental: {
>     jsconfigPaths: true, // enables it for both jsconfig.json and tsconfig.json
>   }
> }
> ```
> 
> Fixes #7779

Hi, I'm new to webpack and next.js. 
How do I add the jsconfigPaths flag if my next.config.js looks like this:

`module.exports = withLess({
   withCSS({
      webpack(config) {
         return config
      }
   })
})`

**Comment by Timer** [comment]
This feature does not require the flag, as it has been released as stable. It should work out of the box.

**Comment by FBosler** [comment]
The aliasing works perfectly for me. Thx for the feature. However, I can't use "code peek" (i.e. command+click on the alias to get to the module) any longer, I am using visual studio code. Any idea why that might be?

Cheers

**Comment by chris-tse** [comment]
Hello, I am using Next 9.4.0 with TypeScript. Setting up the aliases seems to work fine, but Next fails to import the files specified unless the aliases have a slash after the @ symbol (e.g. `import { bla } from @/components` will work, but `@components` will fail). 

Any ideas on why this might be? Could it be attempting to look for a node module with an @ in front, similar to `@angular/core` for instance?

**Comment by lfades** [comment]
@chris-tse I've seen that working in multiple projects, try this:

```json
"paths": {
  "@/*": ["/*"]
}
```

**Comment by chris-tse** [comment]
@lfades I think I figured out the problem. I created a blank tsconfig.json, ran `next dev` to auto-populate the file, then added the paths to it. It wasn't recognizing the paths properly until I closed and restarted the dev server.
